js/pathfinding.js

```diff
@@ -55,27 +53,51 @@ export function findPath(from, to, token, previousWaypoints) {
 	}
 }
 
-function getNode(pos, token, initialize=true) {
-	if (!cachedNodes)
-		cachedNodes = new Array(gridHeight);
-	if (!cachedNodes[pos.y])
-		cachedNodes[pos.y] = new Array(gridWidth);
-	if (!cachedNodes[pos.y][pos.x]) {
-		cachedNodes[pos.y][pos.x] = pos;
+/**
+ * Build a cache ID based on the current token's data and then retrieve the cache to use from that
+ */
+function getCachedNodes(token) {
+	const cacheData = {};
+
+	// Different-sized tokens snap to different points on the grid,
+	// so they might follow a different path to other tokens
+	cacheData.tokenWidth = token.width;
+	cacheData.tokenHeight = token.height;
+
+	// If levels is enabled, the token's elevation can affect which walls
+	// they need to worry about
+	if (game.modules.get("levels")?.active) {
+		cacheData.elevation = token.data.elevation;
+	}
+
+	const cacheId = JSON.stringify(cacheData);
+	// Create a cache if we don't already have one
+	if (!cache.has(cacheId)) {
+		const cachedNodes = new Array(gridHeight);
+		for (let y = 0; y < gridHeight; y++) {
+			cachedNodes[y] = new Array(gridWidth);
+			for (let x = 0; x < gridWidth; x++) {
+				cachedNodes[y][x] = {x, y};
+			}
+		}
+		cache.set(cacheId, cachedNodes);
 	}
+	return cache.get(cacheId);
```

Comment:
If there is a cache hit (which will hopefully happen most of the time) the map lookup needs to be done twice: Once to see whether the element exists and once to fetch it. We could reduce this to a single access like this:
```javascript
let selectedCache = cache.get(cacheId);
if (!selectedCache) {
  selectedCache = initializeCache();
  cache.set(cacheId, selectedCache);
}
return selectedCache;
```

Not a biggie performance wise, but it's easy enough, so we might as well do it.

Reply:
Done.